vars/cloudformation.groovy

```diff
@@ -0,0 +1,224 @@
+  config.parameters.each {
+    params << new Parameter().withParameterKey(it.key).withParameterValue(it.value)
+  }
+  cf.createStack(new CreateStackRequest()
+    .withStackName(config.stackName)
+    .withCapabilities('CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM')
+    .withParameters(params)
+    .withTemplateURL(config.templateUrl))
+}
+
+@NonCPS
+def delete(cf, stackName) {
+  if(doesStackExist(cf, stackName)) {
+    waitUntilComplete(cf, config.stackName)
+    cf.deleteStack(new DeleteStackRequest()
+      .withStackName(stackName)
+    )
+  } else {
+    println "ignoring delete since stack ${stackName} does not exist"
+  }
+}
+
+@NonCPS
+def update(cf, config) {
+  if(doesStackExist(cf, config.stackName)) {
+    waitUntilComplete(cf, config.stackName)
+    def request = new UpdateStackRequest()
+      .withStackName(config.stackName)
+      .withParameters(getStackParams(cf, config.stackName, config.parameters))
+      .withCapabilities('CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM')
+    if(config['templateUrl']) {
+      request.withTemplateURL(config.templateUrl)
+    } else {
+      request.withUsePreviousTemplate(true)
+    }
+    cf.updateStack(request)
+  } else {
+    throw new Exception("unable to update stack ${config.stackName} it does not exist")
+  }
+}
+
+@NonCPS
+def getStackParams(cf, stackName, overrideParams) {
+  def stackParams = [:]
+  def stacks = cf.describeStacks(new DescribeStacksRequest().withStackName(stackName)).getStacks()
+  if (!stacks.isEmpty()) {
+    for(Parameter param: stacks.get(0).getParameters()) {
+      stackParams.put(param.getParameterKey(), new Parameter().withParameterKey(param.getParameterKey()).withUsePreviousValue(true))
+    }
+    overrideParams.each {
+      stackParams.put(it.key, new Parameter().withParameterKey(it.key).withParameterValue(it.value))
+    }
+  }
+  println "stack params: ${stackParams.values()}"
+  return stackParams.values()
+}
+
+@NonCPS
+def wait(cf, stackName, successStatus) {
+  def waiter = null
+  switch(successStatus) {
+    case StackStatus.CREATE_COMPLETE:
+      waiter = cf.waiters().stackCreateComplete()
+    break
+    case StackStatus.UPDATE_COMPLETE:
+      waiter = cf.waiters().stackUpdateComplete()
+    break
+    case StackStatus.DELETE_COMPLETE:
+      waiter = cf.waiters().stackDeleteComplete()
+    break
+  }
+  try {
+    Future future = waiter.runAsync(
+      new WaiterParameters<>(new DescribeStacksRequest().withStackName(stackName)),
+      new NoOpWaiterHandler()
+    )
+    future.get(30, TimeUnit.MINUTES);
+    println "Stack: ${stackName} success - ${successStatus}"
+    return true
+   } catch(Exception e) {
+     println "Stack: ${stackName} failed - ${e}"
+     return false
+   }
+}
+
+@NonCPS
+def doesStackExist(cf, stackName) {
+  try {
+    DescribeStacksResult result = cf.describeStacks(new DescribeStacksRequest().withStackName(stackName))
+    return result != null
+  } catch (AmazonCloudFormationException ex) {
+    return false
```

Comment:
can we just update this to check for error message:

This will allow e.g. credentials error to be detected properly, without being masked with 'stack does not exist' error'

```
if(ex.message.contains("does not exist")){
     return false
} else {
    throw ex
}

Reply:
updated